src/solvers/refinement/string_refinement.cpp

```diff
@@ -303,70 +323,44 @@ void string_refinementt::concretize_string(const exprt &expr)
   {
-    string_exprt str=to_string_expr(expr);
-    exprt length=get(str.length());
+    const string_exprt str=to_string_expr(expr);
+    const exprt length=get(str.length());
     exprt content=str.content();
     replace_expr(symbol_resolve, content);
     found_length[content]=length;
-    mp_integer found_length;
-    if(!to_integer(length, found_length))
-    {
-      INVARIANT(
-        found_length.is_long(),
-        string_refinement_invariantt("the length of a string should be a "
-          "long"));
-      INVARIANT(
-        found_length>=0,
-        string_refinement_invariantt("the length of a string should be >= 0"));
-      size_t concretize_limit=found_length.to_long();
-      INVARIANT(
-        concretize_limit<=generator.max_string_length,
-        string_refinement_invariantt("string length must be less than the max "
-          "length"));
-      exprt content_expr=str.content();
-      std::vector<exprt> result;
-
-      if(index_set[str.content()].empty())
-        return;
+    const auto string_length=expr_cast<std::size_t>(length);
+    INVARIANT(
+      string_length<=generator.max_string_length,
+      string_refinement_invariantt("string length must be less than the max "
+        "length"));
+    if(index_set[str.content()].empty())
+      return;
 
-      // Use the last index as the default character value
-      exprt last_concretized=simplify_expr(
-        get(str[minus_exprt(length, from_integer(1, length.type()))]), ns);
-      result.resize(concretize_limit, last_concretized);
+    // Use the last index as the default character value
+    const exprt last_concretized=simplify_expr(
+      get(str[minus_exprt(length, from_integer(1, length.type()))]), ns);
 
-      // Keep track of the indexes for which we have actual values
-      std::set<size_t> initialized;
+    std::map<std::size_t, exprt> map;
 
-      for(const auto &i : index_set[str.content()])
+    for(const auto &i : index_set[str.content()])
+    {
+      const exprt simple_i=simplify_expr(get(i), ns);
+      try
       {
-        mp_integer mp_index;
-        exprt simple_i=simplify_expr(get(i), ns);
-        if(to_integer(simple_i, mp_index) ||
-           mp_index<0 ||
-           mp_index>=concretize_limit)
-        {
-          debug() << "concretize_string: ignoring out of bound index: "
-                  << from_expr(ns, "", simple_i) << eom;
-        }
-        else
-        {
-          // Add an entry in the result vector
-          size_t index=mp_index.to_long();
-          exprt str_i=simplify_expr(str[simple_i], ns);
-          exprt value=simplify_expr(get(str_i), ns);
-          result[index]=value;
-          initialized.insert(index);
-        }
+        const auto index=expr_cast<std::size_t>(simple_i);
+        const exprt str_i=simplify_expr(str[simple_i], ns);
+        const exprt value=simplify_expr(get(str_i), ns);
+        map.emplace(index, value);
+      }
+      catch(const std::bad_cast &)
```

Comment:
We cannot prevent (for now) the underlying solver to give value to arrays at indexes that have no meaning for us (<0 and >max_string_length). These values should be ignored but there is no bug here.
The clean solution would be to convert the expression to a `mp_integer` and then ignore it if it is negative or too big.